crates/solver/src/settlement_rater.rs

```diff
@@ -244,18 +249,198 @@ impl SettlementRating for SettlementRater {
+        );
+        let probability_fail = BigRational::one() - probability_success.clone();
+        // filter out invalid solution where there is no chance of success or where the
+        // success is 100% (because it can't be true, there is always some risk of
+        // failure)
+        ensure!(!probability_success.is_zero(), "success is zero");
+        ensure!(!probability_fail.is_zero(), "fail is zero");
+
+        let payoff_obj_minus_cap = self.payoff(
+            objective.clone() - self.score_cap.clone(),
+            objective.clone(),
+            probability_success.clone(),
+            cost_fail.clone(),
+        );
+        tracing::trace!(?payoff_obj_minus_cap, "Payoff obj minus cap");
+        let payoff_cap = self.payoff(
+            self.score_cap.clone(),
+            objective.clone(),
+            probability_success.clone(),
+            cost_fail.clone(),
+        );
+        tracing::trace!(?payoff_cap, "Payoff cap");
+        let optimal_bid = if payoff_obj_minus_cap >= BigRational::zero()
+            && payoff_cap <= BigRational::zero()
+        {
+            Ok(probability_success * objective.clone() - probability_fail * cost_fail)
+        } else if payoff_obj_minus_cap >= BigRational::zero() && payoff_cap > BigRational::zero() {
+            Ok(objective.clone()
+                - probability_fail / probability_success * (self.score_cap.clone() + cost_fail))
+        } else if payoff_obj_minus_cap < BigRational::zero() && payoff_cap <= BigRational::zero() {
+            Ok(probability_success / probability_fail * self.score_cap.clone() - cost_fail)
+        } else {
+            Err(anyhow!("Invalid bid"))
+        };
+        tracing::trace!(?optimal_bid, "Optimal bid");
+        if let Ok(optimal_bid) = optimal_bid.as_ref() {
+            ensure!(
+                optimal_bid <= &objective,
+                "Optimal score higher than objective value"
+            );
+        }
+        optimal_bid
+    }
+
+    fn payoff(
+        &self,
+        score_reference: BigRational,
+        objective: BigRational,
+        probability_success: BigRational,
+        cost_fail: BigRational,
+    ) -> BigRational {
+        tracing::trace!(
+            ?score_reference,
+            ?objective,
+            ?probability_success,
+            ?cost_fail,
+            "Computing payoff"
+        );
+        let probability_fail = BigRational::one() - probability_success.clone();
+        let payoff_success = min(objective - score_reference.clone(), self.score_cap.clone());
+        tracing::trace!(?payoff_success, "Payoff success");
+        let payoff_fail = -min(score_reference, self.score_cap.clone()) - cost_fail;
+        tracing::trace!(?payoff_fail, "Payoff fail");
+        let payoff_expectation =
+            probability_success * payoff_success + probability_fail * payoff_fail;
+        tracing::trace!(?payoff_expectation, "Payoff expectation");
+        payoff_expectation
+    }
+}
+
+mod tests {
```

Comment:
Where do these test case come from? Has @fhenneke provided them?

I tried to study https://www.notion.so/cownation/Optimal-bidding-strategy-84b4c710466a4c56af9295015308452a and comprehend which case each of the test cases covers (my assessment is that we are testing case 1 three times, case 2 once and case 3/4 never). It would be good to this to the test name somehow and make sure we cover all cases.

Reply:
I constructed tests 2, 3, and 4 to cover cases 1, 2, and 3, respectively. I think they do, see comment below.

Reply:
Do I understand correctly that in the payoff function based approach there doesn't exist an example for case 4 as it's impossible?